l10n_it_fatturapa/models/ir_attachment.py

```diff
@@ -107,7 +107,7 @@ def get_fattura_elettronica_preview(self):
         xsl_path = get_resource_path(
             "l10n_it_fatturapa",
             "data",
-            self.env.company.fatturapa_preview_style,
+            self.company_id.fatturapa_preview_style,
```

Comment:
Visto https://github.com/OCA/l10n-italy/issues/2965#issuecomment-1267093529
`self.company_id` non punta alla company corrente ma a quella usate per creare l'attachment.

Reply:
Si hai ragione, mi son confuso.
L'errore però sussiste perchè (quando multi-company) self.env.company estrae la company impostata di default nel profilo dell'utente ma la stampa deve avvenire con il template relativo all'azienda in uso. 
Sbaglio?

Reply:
> self.env.company estrae la company impostata di default nel profilo dell'utente

Quello è `self.env.user.company_id`. 

`self.env.company` è la company correntemente selezionata. In teoria.

Reply:
Inizialmente lo pensavo anche io ma entrando in debug mi porta quella di default, confermato anche cambiando il default e vedendo i risultati diversi

Però non ho trovato documentazione ufficiale a riguardo

Reply:
I sorgenti sono qui. 

https://github.com/odoo/odoo/blob/062fb69f346873fe05ebb5d429ab3535fd1df5cc/odoo/api.py#L537-L565

prende la prima tra le allowed_company_ids

se non ci sono, il default dell'utente.

Reply:
Purtroppo da _ir.attachment_ non mi permette di accedere al context andando sempre a finire sul default. Quando passa sulle righe segnate da te non tira fuori niente, ritornando subito il `self.user.company_id`


```
self.context.get('allowed_company_ids', [])
Traceback (most recent call last):
  File "/snap/pycharm-community/302/plugins/python-ce/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
AttributeError: 'ir.attachment' object has no attribute 'context'
```

Reply:
Ho approfondito l'argomento e penso che continui ad essere la miglior scleta utilizzare `self.company_id` in quanto non esiste un context.

Questo metodo scatta da [un controller](https://github.com/OCA/l10n-italy/blob/14.0/l10n_it_fatturapa/controllers/main.py#L4) che passa come unico valore l'id dell'attachment da recuperare e stampare. Gli unici altri dati raggiungibili sono quelli del `user_id` dal quale, quando non trova nessun `allowed_company_ids`, tira fuori il default dell'utente

A meno che non mi sia perso dei parametri al trigger del controller, chiamare `self.env.company` porterà sempre e comunque la company default dell'utente e mai quella attiva.

Dunque penso che, selezionare il template relativo alla company che ha generato l'allegato sia corretto, in quanto altrimenti la funzione di selezione template nelle impostazioni perde di utilità e può creare confusione visto che quelli delle companies diverse dalla default non sono mai accessibli.

Inoltre, nell'eventualità in cui mi sbagli, mi chiedo come sarebbe giusto gestire la selezione del template quando vengono selezionate più companies contemporaneamente. Su che base dare piorità ad una rispetto che un'altra?

Fammi sapre che ne pensi!

Reply:
@michelerusti 

> Dunque penso che, selezionare il template relativo alla company che ha generato l'allegato sia corretto, in quanto altrimenti la funzione di selezione template nelle impostazioni perde di utilità e può creare confusione visto che quelli delle companies diverse dalla default non sono mai accessibli.

Sono d'accordo con questo pensiero, anche perché se l'attachment è stato importato per una company X è giusto avere la preview con lo stile settato nella company X.

Reply:
@michelerusti 
La miglior soluzione è capire perché in quel caso non funziona `self.env.user.company_id`. Nella issue non ci sono istruzioni per riprodurre il problema, anche perché c'è scritto "ogni tanto".  Bisogna capire quando succede esattamente, in modo da riuscire a riprodurre il problema, e poi capire perché succede.

@TonyMasciI 
> Sono d'accordo con questo pensiero, anche perché se l'attachment è stato importato per una company X è giusto avere la preview con lo stile settato nella company X.

Questo non è quello segnalato nella issue. Nella issue ho chiesto esplicitamente quale fosse il comportamento desiderato, ci sono due possibilità:
- si usa il template della company selezionata come corrente dall'utente (`self.env.user.company_id`);
- si usa il template della company che ha emesso la fattura a cui è legato l'attachment (`self.company_id`).
Sono due comportamenti diversi. È stato scelto il primo.

Reply:
@TheMule71 
Si, effettivamente c'è stata un po' di confusione durante tutta questa PR e proverò a fare un sunto della situazione e del perchè penso che la proposta sia valida.

L'errore segnalato nella issue deriva proprio dalla confuzione che si crea utilizzando il `self.env.user.company_id`

L'utilizzo (non mio) di "ogni tanto" è motivato dal fatto che la gestione attuale del template crea confusione in quanto, come dicevo prima, il template di risposta è sempre quello di default utente. La descrizione di come replicare il problema mi sembra ci sia

Il problema deriva dall'assenza di un self.env.context all'interno del controller segnato nel mio commento sopra
Quindi, a meno che esista un modo che non sono riuscito a trovare di accedere al context con le _allowed_company_ cercare di recuperare quello è sbagliato dal principio. 

Quindi, andando avanti in questa PR, alla luce di queste scoperte, propongo di utilizzare `self.company_id` diversamente da come è stato fatto attualmente per effettivamente utilizzare dei valori ai quali possiamio accedere e, come condiviso con @TonyMasciI ,  migliorare l'esperienza utente in quanto a coerenza d'uso.

Potrei ricredermi e cambiare idea se ci fosse un modo di accedere alle `allowed_company_ids` dal controller in questione ma non sono riuscito a trovarlo.

Reply:
ll modo corretto è tramite `self.env.user.company_id`. Quello accede agli `allowed_company_ids`, e fa la cosa giusta.
Su carta almeno.
Se in pratica non funziona, bisogna fixare il bug. In sostanza non va cambiato `self.env.user.company_id`, va fatto funzionare.

Reply:
Ah ok, allora eventualmente ci sarà da sistemare quella parte di codice!

Ad ogni modo, anche una volta fixato, mi sento di portare avanti l'idea che sia meglio usare il `self.company_id` invece di quello delle allowed_companies anche per gli altri motivi portati sopra. 
Pensi sia accettabile?